controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -589,17 +590,49 @@ func (w *Workload) UpdateKubeProxyImageInfo(ctx context.Context, kcp *controlpla
 	if len(ds.Spec.Template.Spec.Containers) == 0 {
 		return nil
 	}
-	newImageName, err := util.ModifyImageTag(ds.Spec.Template.Spec.Containers[0].Image, kcp.Spec.Version)
+
+	parsedImage, err := reference.ParseNormalizedNamed(ds.Spec.Template.Spec.Containers[0].Image)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse image name")
+
+	}
+
+	// Handle imageRepository.
+	imageRepository := fmt.Sprintf("%s/%s", reference.Domain(parsedImage), reference.Path(parsedImage))
+
+	// Handle imageTag.
+	imageRefTag, ok := parsedImage.(reference.Tagged)
+	if !ok {
+		return errors.Errorf("failed to update kube-proxy daemonset: does not have a valid image tag: %q", ds.Spec.Template.Spec.Containers[0].Image)
+	}
+	fromParsedVersion, err := extractImageVersion(imageRefTag.Tag())
 	if err != nil {
 		return err
 	}
 
-	if ds.Spec.Template.Spec.Containers[0].Image != newImageName {
-		patch := client.MergeFrom(ds.DeepCopy())
-		ds.Spec.Template.Spec.Containers[0].Image = newImageName
-		if err := w.Client.Patch(ctx, ds, patch); err != nil {
-			return errors.Wrap(err, "error patching kube-proxy DaemonSet")
-		}
+	toParsedVersion, err := extractImageVersion(kcp.Spec.Version)
+	if err != nil {
+		return err
+	}
+
+	if fromParsedVersion == toParsedVersion {
+		return nil
+	}
+
+	newImageName := fmt.Sprintf("%s:%s", imageRepository, toParsedVersion)
+
+	patch := client.MergeFrom(ds.DeepCopy())
+	ds.Spec.Template.Spec.Containers[0].Image = newImageName
```

Comment:
Potential issue for followup: This should likely iterate over the containers similar to the coreDNS implementation.